packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/textFeatures.ts

```diff
@@ -0,0 +1,152 @@
+        editor.isFeatureEnabled(ExperimentalFeatures.TabKeyTextFeatures) &&
+        !event.rawEvent.shiftKey &&
+        !editor.getElementAtCursor('LI,TABLE', null /*startFrom*/, event),
+    handleEvent: (event, editor) => {
+        const selection = editor.getSelectionRangeEx();
+        if (selection.type == SelectionRangeTypes.Normal) {
+            editor.addUndoSnapshot(() => {
+                if (selection.areAllCollapsed) {
+                    insertTab(editor, event);
+                } else {
+                    if (isWholeParagraphSelected(editor)) {
+                        setIndentation(editor, Indentation.Increase);
+                    } else {
+                        const { ranges } = selection;
+                        const range = ranges[0];
+                        const tempRange = createRange(range.startContainer, range.startOffset);
+
+                        ranges.forEach(range => range.deleteContents());
+                        editor.select(tempRange);
+                        insertTab(editor, event);
+                    }
+                }
+            });
+
+            event.rawEvent.preventDefault();
+        }
+    },
+};
+
+/**
+ * @internal
+ */
+export const TextFeatures: Record<
+    keyof TextFeatureSettings,
+    BuildInEditFeature<PluginKeyboardEvent>
+> = {
+    indentWhenTabText: IndentWhenTabText,
+};
+
+function isWholeParagraphSelected(editor: IEditor): boolean {
+    const regions = editor.getSelectedRegions();
+    let endPosition: NodePosition = null;
+    let startPosition: NodePosition = null;
+
+    regions.forEach(r => {
+        endPosition = r.fullSelectionEnd;
+        startPosition = startPosition || r.fullSelectionStart;
+    });
+
+    const isAtStart: boolean = checkIfIsAtStart(editor, startPosition);
+    const isAtEnd: boolean = checkIfIsAtEnd(editor, endPosition);
+
+    return isAtEnd && isAtStart;
+}
+
+function checkIfIsAtEnd(editor: IEditor, endPosition: NodePosition): boolean {
+    let isAtEnd: boolean;
+    let blockElement = editor.getBlockElementAtNode(endPosition.node).collapseToSingleElement();
+
+    let tempChild = blockElement as ChildNode;
+    while (tempChild.lastChild) {
+        tempChild = tempChild.lastChild ?? null;
+    }
+
+    if (tempChild == endPosition.node && endPosition.isAtEnd) {
+        isAtEnd = true;
+    }
+    return isAtEnd;
+}
+
+function checkIfIsAtStart(editor: IEditor, startPosition: NodePosition): boolean {
+    let isAtStart: boolean;
+    const blockElement = editor.getBlockElementAtNode(startPosition.node).collapseToSingleElement();
+    let tempChild = blockElement.firstChild || blockElement;
+    while (tempChild.firstChild) {
+        tempChild = tempChild.firstChild ?? null;
+    }
+
+    tempChild = ignoreEmptySpans(tempChild);
+
+    if (tempChild == startPosition.node && startPosition.offset == 0) {
+        isAtStart = true;
+    }
+    return isAtStart;
+}
+
+function ignoreEmptySpans(tempChild: ChildNode) {
```

Comment:
this makes me feel the code knows too much detail of the DOM tree, and this check is easy to fail if the tag is not span but something like B, I, U, ...

Can we try this:

Create a range from the real beginning of this block, to the selected position. Then if this range is empty (no text content, no image, no table, no list), then we just treat it as empty, treat the selection is at beginning of block? You can do the same for the end of the block.